### ISSUE TITLE
iScroll vs. native scrolling in 2015

### ISSUE THREAD

**Issue by dt1973** [cubiq/iscroll#976]
Hi!

Is iScroll still needed in 2015 for mobile web / PhoneGap? Doesn't the majority of mobile users now use modern touch devices that support overflow / momentum scrolling?

Thanks!

**Comment by kaansoral** [comment]
I also ask myself the same question every 6 months or so

As far as my assessment goes, iScroll is still needed

I don't know whether there are some hidden features of native scrolling, but my research came empty for basic stuff like:
- determine scroll position
- scroll to end (like reversed scrolling, for messages)

(This is shameful, but, I could never even get the native scrolling to work yet, probably an event handler is preventing it, I've tried it on 5-6 different occasions and moved on with iScroll instead - this thread triggered me to re-asses tho)

**Comment by dt1973** [comment]
:+1: 

Thanks @kaansoral!

**Comment by dt1973** [comment]
> determine scroll position

Can't this be done using simple `height()`, `offset()` etc. calculations?

> scroll to end (like reversed scrolling, for messages)

So assuming one doesn't need reverse scrolling - there is little to no reason to keep on using iScroll?

Curious what the rest of you think as well?

**Comment by kaansoral** [comment]
This properly answers your question by the way: http://blog.ionic.io/native-scrolling-in-ionic-a-tale-in-rhyme/

TL;DR: It still doesn't work, probably in 2016 it will start working diversely

Otherwise I was going to re-visit things today, it's sad that I won't get anywhere

**Comment by dt1973** [comment]
Thanks for sharing. Also see [FTScroller vs. native scrolling in 2015 #103](https://github.com/ftlabs/ftscroller/issues/103).

**Comment by jhnns** [comment]
@dt1973 It depends on what you're trying to achieve. If you "just" want a vertically scrollable list, you can and probably should use native scrolling were possible. The UX is way better (though iScroll is doing a really decent job here).

However, if you need more features like `snap`, etc. or you want an extended API to interact with the element, you're probably better off with iScroll.

Btw, this my code to determine whether to use iScroll or native scrolling (you will need to adjust your UA sniffing/feature detection). Please note: Checking for `WebkitOverflowScrolling` is not enough, because with native scrolling on Android < `4.4`, `click` events on initially hidden elements won't be emitted anymore. Yes... it's true...

```javascript

var hasNativeScrolling = "WebkitOverflowScrolling" in document.documentElement.style ||
    (env.ua.family.search("Chrome") !== -1 && parseInt(env.ua.major, 10) > 25);
var isCrappyAndroid = env.os.family.toLowerCase() === "android" && Number(env.os.major) < 5 && Number(env.os.minor) < 4;
var shouldUseNativeScroll = hasNativeScrolling && isCrappyAndroid === false;

function noop() {}



if (shouldUseNativeScroll) {
    el.style.WebkitOverflowScrolling = "touch";
    el.style.overflowY = "scroll";
    scroll = {
        scrollTo: function (x, y) {
            el.scrollTop = y;
            el.scrollLeft = x;
            // Switching display mode and triggering a synchronous reflow via reading offsetHeight
            // solves rendering issues in modern Chrome and Safari
            el.style.display = "none";
            el.offsetHeight;
            el.style.display = "";
            // Some browsers (Android Stock Browser ~4.1) forget scrolling positions when display = "none" was used
            if (el.scrollTop !== y || el.scrollLeft !== x) {
                el.scrollTop = y;
                el.scrollLeft = x;
            }

        },
        destroy: noop,
        refresh: noop,
        on: noop
    };
    Object.defineProperty(scroll, "maxScrollY", {
        get: function () {
            return el.scrollHeight;
        }
    });
    Object.defineProperty(scroll, "maxScrollX", {
        get: function () {
            return el.scrollWidth;
        }
    });
} else {
    // Fixes bug in Android's WebView where the scroller element just disappeared
    // TODO remove this when iScroll or Webkit has been fixed
    el.firstElementChild.style.WebkitTransitionProperty = "transform";
    scroll = new Scroll(el, {
        scrollX: false,
        scrollY: true,
        tap: "click"
    });
}


scroll.on("beforeScrollStart", function () {
    scroll.refresh();
});
```

**Comment by kaansoral** [comment]
@jhnns thanks A LOT for sharing this, can't wait to test it

(I will probably try listening to the touch events and implementing a basic native/onscroll event that queries .scrollTop, as I have a custom infinite scroll implementation that needs to know when the end is near)
